include/trial/protocol/json/partial/skip.hpp

```diff
@@ -0,0 +1,91 @@
+#ifndef TRIAL_PROTOCOL_JSON_PARTIAL_SKIP_HPP
+#define TRIAL_PROTOCOL_JSON_PARTIAL_SKIP_HPP
+
+///////////////////////////////////////////////////////////////////////////////
+//
+// Copyright (C) 2018 Vinícius dos Santos Oliveira <vini.ipsmaker@gmail.com>
+//
+// Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE_1_0.txt or copy at
+//          http://www.boost.org/LICENSE_1_0.txt)
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include <trial/protocol/json/reader.hpp>
+
+namespace trial
+{
+namespace protocol
+{
+namespace json
+{
+namespace partial
+{
+
+template<class CharT>
+typename basic_reader<CharT>::view_type
+skip(basic_reader<CharT> &reader, std::error_code &ec)
+{
+    using view_type = typename basic_reader<CharT>::view_type;
+    using size_type = typename view_type::size_type;
+
+    switch (reader.symbol()) {
+    case token::symbol::end:
+    case token::symbol::error:
+        ec = errc::insufficient_tokens;
+        break;
+    case token::symbol::null:
+    case token::symbol::boolean:
+    case token::symbol::integer:
+    case token::symbol::real:
+    case token::symbol::string:
+        {
+            auto ret = reader.literal();
+            if (!reader.next())
+                ec = reader.error();
+            return ret;
+        }
+    case token::symbol::begin_array:
+    case token::symbol::begin_object:
+        {
+            auto ret_begin = reader.literal().data();
+            const auto outer_level = reader.level();
+            view_type last_token;
+            do {
+                last_token = reader.literal();
+                if (!reader.next()) {
+                    if (reader.level() != outer_level)
+                        ec = errc::insufficient_tokens;
+                    else
+                        ec = reader.error();
+                    break;
+                }
+            } while (reader.level() != outer_level);
+            auto ret_end = last_token.data() + last_token.size();
+            auto ret_size = ret_end - ret_begin;
+            return {ret_begin, static_cast<size_type>(ret_size)};
+        }
+    case token::symbol::end_array:
+    case token::symbol::end_object:
+        assert(false);
```

Comment:
The user can call `skip` when `reader` is pointing to the end of an array/object, so it may be better to either

1. Skip over the token.
2. Report an error.

Reply:
Whoops, you're right. The restrictions are different now that I'm making a function to the general user.

Reply:
It was hard to come up with a compelling argument to choose either `#1` or `#2`. Then I finally thought about something that pleases me: `partial::skip` and `partial::parse` should be interchangeable. I guess `partial::parse` will error out on `}` because that is an invalid JSON value, then `partial::skip` should follow this behaviour as well.

Reply:
I agree.